### PR TITLE
Automatically retrieve temporary S3 credentials on AWS instances.

### PR DESCRIPTION
**Pull request by dwsteele**
Automatically retrieve temporary credentials for S3 when the AWS instance is associated with an IAM role. Credentials are automatically updated when they are <= 5 minutes from expiring.

Basic configuration is to set `repo1-s3-key-type=auto`. `repo1-s3-role` can be used to set a specific role, otherwise it will be retrieved automatically. We welcome discussion of these option names/values.

Most of the changed files deal with configuration -- the main code changes are in `src/storage/s3/storage.c`.

### CODE REVIEW

**Comment by dwsteele** [comment]
@Xof, you have expressed interest in this feature in the past. We'd appreciate it if you could have a look, doing some testing, comment on the functionality, etc.

File: src/storage/s3/storage.c [hunk 1]
```diff
@@ -273,6 +315,99 @@ storageS3RequestAsync(StorageS3 *this, const String *verb, const String *uri, St
...
+                httpRequestError(request, response);
+
+            // Free old credentials
+            strFree(this->accessKey);
+            strFree(this->secretAccessKey);
+            strFree(this->securityToken);
+
+            // Get credentials from the JSON response
+            KeyValue *credential = jsonToKv(strNewBuf(httpResponseContent(response)));
+            (void)credential;
```
Comment (cmwshang) [inline src/storage/s3/storage.c:380]: Casting to void is often used to suppress warnings for unused variables but I don't see why this is done here in (void)credential; - there are no conditionally compiled code sections with the credential variable so can you explain why this is needed here?

Comment (dwsteele) [reply src/storage/s3/storage.c:380]: Looks I put that in so I could get it to compile before writing the rest of the section. Removed.

**Comment by youattd** [comment]
Built and tested on Amazon Linux v1 and v2.  
Using CloudTrail you can see the AssumeRole API events happen about every hour (default).
In the AssumeRole event body, the new temp key has an expiration time about 6 (or 8?) hours from the event timestamp.

**Comment by aleszeleny** [comment]
Hello,

thanks a lot for the amazing work done!

Based on https://github.com/pgbackrest/pgbackrest/issues/721 I've tried to run a backup having:
`repo1-s3-key-type=auto`
and commented out
```
#repo1-s3-key=
#repo1-s3-key-secret=
```

Stanza creation proceeded, `info` command for other stanzas having some backups in place worked as well.

The backup is running (testing on ~1TB pg cluster, so it'll take some time to finish (dev server only 2 parallel processes).

I doubt the credential updates "<= 5 minutes from expiring", there are many cases (like `vacuumdb` on an active instance running 4 parallel jobs causing pgBackRest (async push, 4 parallel processes on 16vCPU machine) archiving thousands (30k...) of WAL files (default size 16MB each) takes 20-40 minutes to complete (server was not overloaded, so adding more parallel push processes is an option). From Credential timing perspective, it depends whether pgBackrest use somewhat cashed credentials and if all WAL files are sent using single HTTP stream to S3 (therefore once authentication is done, the credentials might expire with no issue) or if it opens a new session to S3 for every WAL file using the cached credentials ( and it'll be an issue unless retrieval fo new credential will be triggered).

I'll update you the backup results, it'll take some more time to prepare an environment where to restore from the testing backup.

Once again, thanks for that change!
Ales

**Comment by dwsteele** [comment]
@aleszeleny Yes -- we don't reuse credentials between executions of pgBackRest. To do so we'd need to cache them somewhere and for complexity/security reasons it seems better just to request them again. So renewal is just for longer running session that have credentials that will expire before the work is done.

**Comment by JBromage** [comment]
Hi, 
I've also tested this against a 5TB db ( what are limits if they aren't to be pushed after all ) - in all it took 15 hours, 34 minutes, but it had been throttled back as I wanted it to potentially fail as it had been doing after 6 hours......I just used the parameter repo1-s3-key-type=temp, still using an instance profile IAM , so normally the temp token would expire after 6 hours.... and without any intervention, its completed without any errors / having to resume the backup. 
This is absolutely what I needed, and thank you for all your effort on this. If you need any details / help with documentation for this, I've got all the logfiles etc so can provide sample outputs and sample config. Thank you again, this is brilliant!

**Comment by aleszeleny** [comment]
Hello, my testing backup succeeded as well:
```2020-08-25 04:58:57.597 P00   INFO: full backup size = 2405.4GB
2020-08-25 04:58:57.597 P00   INFO: execute non-exclusive pg_stop_backup() and wait for all WAL segments to archive
2020-08-25 04:58:58.098 P00   INFO: backup stop archive = 00000002000058790000003B, lsn = 5879/3B97FB28
2020-08-25 04:58:58.262 P00   INFO: check archive for segment(s) 000000020000587900000014:00000002000058790000003B
2020-08-25 04:58:59.950 P00   INFO: new backup label = 20200824-123648F
2020-08-25 04:59:00.970 P00   INFO: backup command end: completed successfully (58852039ms)
```
Thanks for the fantastic feature!

**Comment by dwsteele** [comment]
@JBromage I have added the user guide documentation if you would like to have a look: https://github.com/pgbackrest/pgbackrest/pull/1137/files#diff-a62458990193106959efca32e11bc8ef

I'd rather not provide specific AWS-side configuration in this case since I don't think we need to cover that level of detail in the pgBackRest user guide.

**Comment by JBromage** [comment]
> @JBromage I have added the user guide documentation if you would like to have a look: https://github.com/pgbackrest/pgbackrest/pull/1137/files#diff-a62458990193106959efca32e11bc8ef
> 
@dwsteele - yep, that all looks good to me  - -thank you again for this

**Comment by dwsteele** [comment]
And committed! Thank you to everyone for your assistance reviewing and testing this patch.

**Pull request merged**